pep-0722.rst

```diff
@@ -17,87 +17,136 @@ Resolution:
-other input type(s).
-["the other input type(s)" -- There's only one input type. It's a Union. Suggest rephrasing this. I'm not sure if talking about the types using set theory (input -- output) would make this more clear (or more generic) or worse.]
+Since the ``TypeGuard`` type (or output type) is a subtype of the input argument
+type, a type checker can determine that the only possible type in the ``else``
+is the other type in the Union. In this example, it is safe to assume that if
+``is_str`` returns false, then type of the ``val`` argument is an ``int``.
 
-If the output type is not a *strict* subtype of the input type,
-the negative cannot be assumed to be the intuitive opposite:
-["intuitive opposite" -- opposite is the incorrect term here and I think intuition doesn't belong in a PEP :)]
+Unsafe Narrowing
+--------------------
 
-::
+There are cases where this further type narrowing is not possible. Here's an
+example:
+
+.. code-block:: python
 
     def is_str_list(val: list[int | str]) -> TypeGuard[list[str]]
         return all(isinstance(x, str) for x in val)
 
     def func(val: list[int | str]):
         if is_str_list(val):
-            reveal_type(val) # list[str]
+            # Type checker assumes list[str] here
+        else:
+            # Type checker cannot assume list[int] here
+
+Since ``list`` is invariant, it doesn't have any subtypes. This means type
+checkers cannot narrow the type to ``list[int]`` in the false case.
+``list[str]`` is not a subtype of ``list[str | int]``.
+
+Type checkers should not assume any narrowing in the false case when the
+``TypeGuard`` type is not a subtype of the input argument type. 
+
+However, narrowing in the true case is still possible. In the example above, the
+type checker can assume the list is a ``list[str]`` if the ``TypeGuard``
+function returns true.
+
+Creating invalid narrowing
+--------------------------
+
+The new ``else`` case for a ``TypeGuard`` can be setup incorrectly. Here's an example:
+
+.. code-block:: python
+
+    def is_positive_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes int here
+        else:
+            # Type checker assumes str here
+
+A type checker will assume for the else case that the value is ``str``. This
+is a change in behavior from :pep:`647` but as that pep stated `here <https://peps.python.org/pep-0647/#enforcing-strict-narrowing>`__
+there are many ways a determined or uninformed developer can subvert type safety.
+
+A better way to handle this example would be something like so:
+
+.. code-block:: python
+
+    PosInt = NewType('PosInt', int)
+
+    def is_positive_int(val: PosInt | int | str) -> TypeGuard[PosInt]:
+        return isinstance(val, int) and val > 0
+
+    def func(val: int | str):
+        if is_positive_int(val):
+            # Type checker assumes PosInt here
```

Comment:
> Type checker assumes PosInt here

This is true even though `val` is `int | str` (i.e. it doesn't include `PosInt`)?

Reply:
Yes that's true. This works because PosInt is an int.

Reply:
You should be able to try it yourself in the latest pylance.

Create a pyrightconfig.json with this in it:

```json
{
    "enableExperimentalFeatures" : true
}
```

Then the TypeGuard will behave as I described here.

Reply:
I believe that that's how Pyright works. I guess I'm wondering:

a) Is this correct?

b) What would be the behavior if `PosInt` was unrelated to any of the types in the `val` union? Consider the following example. Within the `if` block, is `val` considered to be `datetime` because `is_datetime` returned `true`? Or is it `Never` because the intersection of `datetime` and `int | str` is empty? Oh, I guess this might be the multiple inheritance thing again. Maybe the `val` passed into `func` is derived from both `int` and `datetime`?

```python
def is_datetime(val: datetime | int | str) -> TypeGuard[datetime]:
    return isinstance(val, datetime)

def func(val: int | str):
    if is_datetime(val):
        # Is val a datetime here or Never?
```

c) Should the PEP call more attention to these issues?

Reply:
I believe the TypeGuard pep says that if the TypeGuard returns something nonsensical, that's the user's fault. It will just happily presume that's the actual type if the TypeGuard returns true.

This is exactly what Pyright does at the moment.

Reply:
It's mentioned here:
https://peps.python.org/pep-0647/#enforcing-strict-narrowing